Add VI Analyzer test coverage to PRs

aepete opened the pull request:
<!-- TODO: Mark the following with an 'x' as applicable -->
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md). _(Required)_
- [x] <!--G_DIFF_CHECK--> Automatically post PR comments with images for G code changes? _(Recommended for small changes)_

### What does this Pull Request accomplish?

Exercises VI Analyzer via a workflow on PRs.

### Why should this Pull Request be merged?

VI Analyzer helps ensure quality LabVIEW code, ranging from correct behavior to performance to style & readability.

_Important_: This change introduces VI Analyzer tooling while being relatively non-invasive to the pre-existing source code (.., and the numerous pre-existing issues that a full VI Analyzer run would flag!)  Therefore, this initial implementation only exercises VI Analyzer for limited cases: specifically, it covers the G contents under the `Runtime` folder for most level-2 (high-severity) cases covered by VI Analyzer.  Ideally this tooling shall evolve to cover more of our source code (perhaps under `Generator`, `Tests`, or `Examples`) and validate more test cases (most importantly, the remaining level-2 cases... then also level-1 & level-0).

The G changes in this PR were made in order to address the pre-existing high-severity issues that are covered by `PR.viancfg`

In order to run this testing on runner `spawningpool`, I've installed VI Analyzer Toolkit 2020 64-bit via NI Package Manager.

### What testing has been done?

Exercised code locally and within PR workflow.

Validated the failure case properly reports failure text to the log.

Validated the success case properly reports progress text to the log and the workflow passes.

aepete commented on Source/Tools/PR.viancfg:
```diff
@@ -0,0 +1,1062 @@
+		<MaxFailures>5</MaxFailures>
+		<BasePath>"LabVIEW"</BasePath>
+		<RelativePath>"project\\_VI Analyzer\\_tests\\Block Diagram\\Style\\Unused Code.llb"</RelativePath>
+		<Selected>FALSE</Selected>
```
This VI Analyzer configuration is set to run most tests with a default `<Ranking>2</Ranking>` (high severity) and not run tests with `<Ranking>1</Ranking>` (medium severity) or `<Ranking>0</Ranking>` (low severity).  The only exceptions are three disabled high-severity tests:
- `Unused Code` disabled due to some of grpc-labview's generated code having unused unbundle nodes
- `Broken VI` disabled due to a handful of flagged cases.
  - We have a G test to cover broken VIs too.  However, perhaps we should strive to enable the broken VI coverage via VI Analyzer, then perhaps the G test could be removed..?
- `Reentrant VI Issues` disabled due to a number of flagged cases, most of which involve a non-reentrant subVI being called from a reentrant caller.

pbirkhol-ni replied:
I'm surprised that VI Analyzer caught broken VIs but our test didn't. In any case, I agree we should rely on VI Analyzer and not our test. However, we should do that as part of another PR.

aepete replied:
I think the problem actually may have been that I didn't have all the drivers on my local develoment setup.  Maybe the check would work when run on spawningpool.

aepete replied:
I'll try it out in a follow-up PR :)

pbirkhol-ni replied:
Sounds good. We do our best to make sure nothing is broken, even if a driver isn't installed. However, there is one private VISA class VI that will be broken if you don't have VISA 24.0 installed. Since this breakage never propagates to customers and since there isn't an ideal fix, we left it as is. But it would definitely trip up VI analyzer.

dixonjoel commented on .github/workflows/pr.yml:
```diff
@@ -15,6 +15,11 @@ jobs:
 
+  run_vi_analyzer:
+    name: Run VI Analyzer
+    if: ${{ github.repository == 'ni/measurementlink-labview' || github.event.pull_request.head.repo.full_name == 'ni/measurementlink-labview' }}
```
I see this same 'if' check is done above, but do you know why we need this? Isn't the PR always run on 'this' repo, which is always ni/measurementlink-labview?

aepete commented:
It's possible that @pbirkhol-ni's #457 addresses some issues that would be caught by level-1 or level-0 checks in VIAnalyzer.

Meanwhile I'll note that since #457 touches a lot of VIs, we'll want to be cautious of merge issues.  Fortunately, the G changes in this PR (#458) are quite straightforward and could be re-implemented easily if need be.

pbirkhol-ni commented:
> It's possible that @pbirkhol-ni's #457 addresses some issues that would be caught by level-1 or level-0 checks in VIAnalyzer.
> 
> Meanwhile I'll note that since #457 touches a lot of VIs, we'll want to be cautious of merge issues. Fortunately, the G changes in this PR (#458) are quite straightforward and could be re-implemented easily if need be.

I will merge if necessary. I doubt my change will address all of the level 1 or level 0 issues, but maybe some of them. If I have time next week, I might start addressing more issues. I also hope VI Analyzer will allow us to run customer code that checks for issues.

aepete commented on Source/Tools/run_diff.py:
```diff
@@ -109,7 +109,7 @@ def get_changed_labview_file_paths(repo_root_directory, pr_number, token):
         changed_files = get_git_diff_changed_files()
 
-    diff_regex = re.compile(r"^(.*\.vi[tm]?)|(.*\.ctl?)$", re.MULTILINE)
+    diff_regex = re.compile(r"^.*\.(vi|vit|vim|ctl)$", re.MULTILINE)
```
I updated this regex because the old match was considering VI Analyzer config files (`.viancfg`) a match, thus would've otherwise gone down the path of producing diff imagery.

pbirkhol-ni commented on Source/Runtime/MeasurementLink Discovery V1/Service Location FGV Impl.vi:
Can you elaborate on how this VI changed? Same with all of the other VIs that changed without a diff. The text above says it is the VI Icon, but I can't spot the difference.  :)

pbirkhol-ni replied:
Nevermind. I don't have your branch checked out. Let me try that first.

pbirkhol-ni replied:
Okay. I have it checked out and still can't see a difference in the icon.

aepete replied:
In these cosmetic-only / no-image cases, I had to move a nested case structure or loop or constant so there were zero pixels of overlap with the owning structure.  None were very severe to begin with, but VIAnalyzer considers any pixel of overlap as level-2.

aepete replied:
FTR, I have a backlog item to allow for the graphical diff to have an opt-in to include cosmetic changes.  That feature would've been helpful in this case.

pbirkhol-ni replied:
Okay. That makes sense. I will likely have to make similar changes to fix merge conflicts so it is good to know the issue.

pbirkhol-ni commented on .github/workflows/pr.yml:
```diff
@@ -15,6 +15,11 @@ jobs:
     if: ${{ github.repository == 'ni/measurementlink-labview' || github.event.pull_request.head.repo.full_name == 'ni/measurementlink-labview' }}
     uses: ./.github/workflows/run_g_tests.yml
 
+  run_vi_analyzer:
```
@aepete, we're eventually going to want to run these tests (as well as the differ) on measlinkbot. Can you let @MounikaBattu17 what extra software needs to be installed to run VI Analyzer and the differ?

aepete replied:
Diff check requires no additional tooling -- just LabVIEW.

VI Analyzer check requires **VI Analyzer Toolkit 2020 64-bit**

pbirkhol-ni commented on Source/Tools/run_vi_analyzer.py:
```diff
@@ -0,0 +1,38 @@
+    _logger.debug(formatted_stdout)
+    if(test_result.returncode != 0):
+        formatted_stderr = test_result.stderr.decode().replace('\r\n','\n').strip()
+        _logger.error(formatted_stderr)
```
Is this what we would look at to debug any failures? Do you have an example of what this looks like?

aepete replied:
Yes!  Shown below is an example of failures pertaining to overlapping diagram content, which will be viewable from the workflow failure log:

```
VI Analyzer Results

Original Analysis Performed	Wednesday, December 13, 2023 4:08:00 PM
Total Analysis Time	00:02:53.22

Results
VIs Analyzed	779
Total Tests Run	13243
Passed Tests	13242
Failed Tests	1
Skipped Tests	0

Errors
VI not loadable	0
Test not loadable	0
Test not runnable	0
Test error out	0

Failed Tests (sorted by VI)

Start Discovery Service.vi (C:\github\measurementlink-labview\Source\Runtime\MeasurementLink Discovery V1\Start Discovery Service.vi)
Hidden Objects in Structures	This constant is not entirely within the visible bounds of its owner (Case Structure, "False" case).
Hidden Objects in Structures	This constant is not entirely within the visible bounds of its owner (Case Structure, "False" case).
Hidden Objects in Structures	The subVI "Error Cluster From Error Code.vi" is not entirely within the visible bounds of its owner (Case Structure, "False" case).
Hidden Objects in Structures	This global variable (5056) is not entirely within the visible bounds of its owner (Case Structure, "False" case).
Hidden Objects in Structures	This constant is not entirely within the visible bounds of its owner (Case Structure, "False" case).


Testing Errors

(none)
```

pbirkhol-ni commented on Source/Tools/PR.viancfg:
I'm not going to review this entire file. I assume it is the standard VI Analyzer file.

aepete commented on Source/Tools/run_vi_analyzer.py:
```diff
@@ -0,0 +1,38 @@
+    return test_result.returncode
+
+
+main()
```
I'll clean up this case and the other workflow-induced python scripts in a separate change.

aepete replied:
I've created a work item for this.
